File: src/engine/pairPool.ts
import { pairKey } from "../content/schema";
import type { Pack, WordPair } from "../content/types";
import { pickIndex, type Rng } from "./rng";

export type PairPool = {
  pairs: readonly WordPair[];
  usedKeys: ReadonlySet<string>;
};

export type DrawResult =
  | { kind: "ok"; pair: WordPair; pool: PairPool }
  | { kind: "exhausted" };

export function buildPool(packs: readonly Pack[]): PairPool {
  return {
    pairs: packs.flatMap((pack) => pack.pairs),
    usedKeys: new Set<string>(),
  };
}

function availablePairs(pool: PairPool): WordPair[] {
  return pool.pairs.filter((pair) => !pool.usedKeys.has(pairKey(pair)));
}

export function remainingCount(pool: PairPool): number {
  return availablePairs(pool).length;
}

/** Exhaustion is a returned value, never a throw — the UI offers recycling. */
export function drawPair(pool: PairPool, rng: Rng): DrawResult {
  const available = availablePairs(pool);
  if (available.length === 0) return { kind: "exhausted" };

  const pair = available[pickIndex(available.length, rng)]!;
  const usedKeys = new Set(pool.usedKeys);
  usedKeys.add(pairKey(pair));

  return { kind: "ok", pair, pool: { pairs: pool.pairs, usedKeys } };
}

export function recyclePool(pool: PairPool): PairPool {
  return { pairs: pool.pairs, usedKeys: new Set<string>() };
}

File: flutter_app/lib/engine/pair_pool.dart
import '../content/types.dart';
import 'rng.dart';

class PairPool {
  final List<WordPair> pairs;
  final Set<String> usedKeys;

  const PairPool({required this.pairs, required this.usedKeys});

  factory PairPool.build(Iterable<Pack> packs) => PairPool(
        pairs: [for (final pack in packs) ...pack.pairs],
        usedKeys: const {},
      );

  List<WordPair> get _available =>
      pairs.where((p) => !usedKeys.contains(p.key)).toList();

  int get remaining => _available.length;

  PairPool get recycled => PairPool(pairs: pairs, usedKeys: const {});
}

/// Exhaustion is a returned value, never a throw — the UI offers recycling.
sealed class DrawResult {
  const DrawResult();
}

class DrawOk extends DrawResult {
  final WordPair pair;
  final PairPool pool;
  const DrawOk(this.pair, this.pool);
}

class DrawExhausted extends DrawResult {
  const DrawExhausted();
}

DrawResult drawPair(PairPool pool, Rng rng) {
  final available = pool._available;
  if (available.isEmpty) return const DrawExhausted();

  final pair = available[rng.pickIndex(available.length)];
  return DrawOk(
    pair,
    PairPool(pairs: pool.pairs, usedKeys: {...pool.usedKeys, pair.key}),
  );
}
